systemservices/marketplace/src/tasks/preparePurchase.ts

```diff
@@ -13,12 +13,26 @@ export default (
   const transactions: Promise<any>[] = []
   let shiftNonce = 0
   try {
+    // get service
+    const service = await getService(marketplace, inputs.sid)
+    
+    // check ownership
+    if (service.owner.toLowerCase() !== inputs.from.toLowerCase()) throw new Error(`service's owner is different that the specified 'from'`)
+
     // get offer data
-    const offer = await getServiceOffer(marketplace, inputs.sid, new BigNumber(inputs.offerIndex))
+    const offerIndex = new BigNumber(inputs.offerIndex).toNumber()
```

Comment:
should check for negative numbers as well

Reply:
It's checked at the next line

Reply:
it only checks for bigger indexes but how about `...-2, -1`?

Reply:
secondly, _toNumber()_ can return a float value number too. we need to make sure that we're using integer inside arrays. gotto check this in other similar places too. there is `integerValue ()` func inside `BigNumber` maybe it can suit here.

Reply:
it's already done: https://github.com/mesg-foundation/core/pull/874/files#diff-ad93658a1e6d166abbf4848e45444a02R25

Reply:
oh okay, I was reviewing the first commit related to this, didn't see this update yet ;) but we need to make sure to use integers for the second one.

Reply:
Done in https://github.com/mesg-foundation/core/compare/ss/marketplace-improve-checks-more?expand=1